public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -133,16 +97,196 @@ export function usePauseRuleInGroup() {
+      }
+
+      const action = moveRuleGroupAction({ namespaceName, groupName, interval });
+      const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);
+
+      const oldNamespace = ruleGroup.namespaceName;
+      const targetNamespace = action.payload.namespaceName;
+
+      const oldGroupName = ruleGroup.groupName;
+      const targetGroupName = action.payload.groupName;
+
+      const isGroupRenamed = Boolean(targetGroupName) && oldGroupName !== targetGroupName;
+
+      // if we're also renaming the group, check if the target does not already exist
+      if (targetGroupName && isGroupRenamed) {
+        const targetGroup = await fetchRuleGroup({
+          rulerConfig,
+          namespace: targetNamespace,
+          group: targetGroupName,
+        })
+          .unwrap()
+          .catch(notFoundToNull);
+
+        if (targetGroup?.rules?.length) {
+          throw new Error('Target group already has rules, merging rule groups is currently not supported.');
+        }
+      }
+
+      // create the new group in the target namespace
+      await upsertRuleGroup({
+        rulerConfig,
+        namespace: targetNamespace,
+        payload: newRuleGroupDefinition,
+      }).unwrap();
+
+      const result = await deleteRuleGroup({
+        rulerConfig,
+        namespace: oldNamespace,
+        group: oldGroupName,
+      }).unwrap();
+
+      return result;
+    },
+    [produceNewRuleGroup, fetchRuleGroup, upsertRuleGroup, deleteRuleGroup]
+  );
+
+  return [moveFn, state] as const;
+}
+
+/**
+ * Rename a rule group but keep it within the same namespace, throws if the action targets an existing rule group
+ */
+export function useRenameRuleGroup() {
+  const [produceNewRuleGroup] = useProduceNewRuleGroup();
+  const [fetchRuleGroup] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
+  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();
+  const [deleteRuleGroup] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useMutation();
+
+  const [state, renameFn] = useAsyncFn(
+    async (ruleGroup: RuleGroupIdentifier, groupName: string, interval?: string) => {
+      const action = renameRuleGroupAction({ groupName, interval });
+      const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);
+
+      const oldGroupName = ruleGroup.groupName;
+      const newGroupName = action.payload.groupName;
+      const namespaceName = ruleGroup.namespaceName;
+
+      // check if the target group exists
+      const targetGroup = await fetchRuleGroup({
+        rulerConfig,
+        namespace: namespaceName,
+        group: newGroupName,
+      })
+        .unwrap()
+        .catch(notFoundToNull);
+
+      if (targetGroup?.rules?.length) {
+        throw new Error('Target group has existing rules, merging rule groups is currently not supported.');
```

Comment:
This does indeed throw, but it seems to be swallowed elsewhere and the modal just closes without showing a message to the user when you try and rename like this

Reply:
I agree, I think we should try catch this error in the modal, show the error and not close it.

Reply:
This seems to be a regression because `useAsyncFn` [does not rethrow](https://github.com/streamich/react-use/issues/1923) :(

Reply:
I've resolved this by just vendoring a different hook to use for async fns – the interface is slightly different from `react-use/useAsyncFn` but I can live with it :)

Thoughts?

Reply:
but then the `asyncState.error` is the indicator of the thrown exception?

Reply:
It would yes but you wouldn't be able to use that for effects that need to happen before a re-render occurs such as closing a modal or showing a toast. For those use-cases you'd need to be able to either catch the Promise or infer it from the result of the hook.